src/lib/sql/sql_pipeline_statement.cpp

```diff
@@ -105,26 +110,51 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
     return _query_plan;
   }
 
-  const auto& lqp = get_optimized_logical_plan();
+  // If we need a transaction context but haven't passed one in, this is the latest point where we can create it
+  if (!_transaction_context && _use_mvcc) {
+    _transaction_context = TransactionManager::get().new_transaction_context();
+  }
+
   _query_plan = std::make_shared<SQLQueryPlan>();
 
-  const auto started = std::chrono::high_resolution_clock::now();
+  auto started = std::chrono::high_resolution_clock::now();
+
+  // Handle query plan if statement has been cached
+  const auto cached_plan = SQLQueryCache<SQLQueryPlan>::get().try_get(_sql_string);
+  if (cached_plan) {
+    auto& plan = *cached_plan;
+
+    DebugAssert(!plan.tree_roots().empty(), "QueryPlan retrieved from cache is empty.");
+    if (plan.tree_roots().front()->transaction_context_is_set()) {
+      Assert(_use_mvcc, "Cannot switch from MVCC use to non-MVCC use with cached queries.");
+    } else {
+      Assert(!_use_mvcc, "Cannot switch from non-MVCC use to MVCC use with cached queries.");
```

Comment:
It is not clear to me if the cached version has MVCC or not. Maybe: "Trying to use non-MVCC cached query with a transaction context" and "Trying to use MVCC-enabled cached query without ..."?